torch/distributed/_spmd/data_parallel.py

```diff
@@ -0,0 +1,731 @@
+            if node_type == NodeType.NON_TENSOR:
+                # set node sharding to None
+                node_sharding = None
+            elif placeholder_idx < activation_idx:
+                assert len(node_strategies) > 0, "node_strategies should not be empty"
+                if parallel_mode == DataParallelStyle.REPLICATE:
+                    # set to replicate for replicate style
+                    node_sharding = node_strategies[0]
+                elif parallel_mode == DataParallelStyle.FULLY_SHARD:
+                    # set to shard for fully shard style
+                    node_sharding = node_strategies[1]
+                elif parallel_mode == DataParallelStyle.DEFAULT:
+                    # todo: add support for default mode
+                    # default mode would generate either replicate or shard
+                    raise NotImplementedError("default mode not implemented")
+            else:
+                assert len(node_strategies) > 0, "node_strategies should not be empty"
+                # mark activation as sharded on batch dim
+                node_sharding = node_strategies[0]
+
+            node.meta["sharding"] = node_sharding
+
+            placeholder_idx += 1
+        elif node.op == "call_function":
+            node_strategies = dp_strategy_map[node].strategies
+            assert (
+                len(node_strategies) <= 2
+            ), "data parallel should have at most 2 strategies"
+            if len(node_strategies) == 1:
+                node.meta["sharding"] = node_strategies[0]
+            elif len(node_strategies) == 2:
+                if parallel_mode == DataParallelStyle.REPLICATE:
+                    # set to replicate for replicate style
+                    node.meta["sharding"] = node_strategies[0]
+                elif parallel_mode == DataParallelStyle.FULLY_SHARD:
+                    # set to shard for fully shard style
+                    node.meta["sharding"] = node_strategies[1]
+                else:
+                    raise RuntimeError("default mode not supported yet!")
+            else:
+                raise RuntimeError(
+                    f"node {node} strategy length {len(node_strategies)} is not expected!"
+                )
+        elif node.op == "output":
+            assert node_type == NodeType.NON_TENSOR, "output node should not be tensor"
+            node.meta["sharding"] = None
+        else:
+            raise RuntimeError(f"op code {node.op} not supported")
+
+
+def _to_local_shard(
+    full_tensor: torch.Tensor, spec: DTensorSpec, from_local: bool = False
+) -> torch.Tensor:
+    """
+    util function to convert a full tensor to its local component
+    """
+    local_shard = full_tensor
+    if from_local:
+        # flag indicate the tensor is already local, we don't need to do anything
+        return local_shard
+
+    for idx, placement in enumerate(spec.placements):
+        if placement.is_shard():
```

Comment:
Curious: Can this only be `True` for one placement in `spec.placements`?

Reply:
not really, we can have something like `[Shard(0), Shard(1)]` on a 2-d mesh

Reply:
I misread the implementation. You continue sharding `local_shard` for each placement. This makes sense now :)